component/loki/source/api/internal/lokipush/push_api_server.go

```diff
@@ -0,0 +1,240 @@
+	"github.com/grafana/agent/component/common/loki"
+	fnet "github.com/grafana/agent/component/common/net"
+	frelabel "github.com/grafana/agent/component/common/relabel"
+	"github.com/grafana/dskit/tenant"
+	"github.com/grafana/loki/pkg/loghttp/push"
+	"github.com/grafana/loki/pkg/logproto"
+	util_log "github.com/grafana/loki/pkg/util/log"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/prometheus/model/labels"
+	"github.com/prometheus/prometheus/model/relabel"
+	promql_parser "github.com/prometheus/prometheus/promql/parser"
+)
+
+type PushAPIServer struct {
+	logger       log.Logger
+	serverConfig *fnet.ServerConfig
+	server       *fnet.TargetServer
+	handler      loki.EntryHandler
+
+	rwMutex       sync.RWMutex
+	labels        model.LabelSet
+	relabelRules  []*relabel.Config
+	keepTimestamp bool
+}
+
+func NewPushAPIServer(logger log.Logger,
+	serverConfig *fnet.ServerConfig,
+	handler loki.EntryHandler,
+	registerer prometheus.Registerer,
+) (*PushAPIServer, error) {
+
+	s := &PushAPIServer{
+		logger:       logger,
+		serverConfig: serverConfig,
+		handler:      handler,
+	}
+
+	srv, err := fnet.NewTargetServer(logger, "agent_loki_source_api", registerer, serverConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	s.server = srv
+	return s, nil
+}
+
+func (s *PushAPIServer) Run() error {
+	level.Info(s.logger).Log("msg", "starting push API server")
+
+	err := s.server.MountAndRun(func(router *mux.Router) {
+		router.Path("/api/v1/push").Methods("POST").Handler(http.HandlerFunc(s.handleLoki))
+		router.Path("/api/v1/raw").Methods("POST").Handler(http.HandlerFunc(s.handlePlaintext))
+		router.Path("/ready").Methods("GET").Handler(http.HandlerFunc(s.ready))
+	})
+	return err
+}
+
+func (s *PushAPIServer) ServerConfig() fnet.ServerConfig {
+	return *s.serverConfig
+}
+
+func (s *PushAPIServer) Shutdown() {
+	level.Info(s.logger).Log("msg", "stopping push API server")
+	s.server.StopAndShutdown()
+}
+
+func (s *PushAPIServer) SetLabels(labels model.LabelSet) {
+	s.rwMutex.Lock()
+	defer s.rwMutex.Unlock()
+	s.labels = labels
+}
+
+func (s *PushAPIServer) getLabels() model.LabelSet {
+	s.rwMutex.RLock()
+	defer s.rwMutex.RUnlock()
+	return s.labels.Clone()
+}
+
+func (s *PushAPIServer) SetKeepTimestamp(keepTimestamp bool) {
+	s.rwMutex.Lock()
+	defer s.rwMutex.Unlock()
+	s.keepTimestamp = keepTimestamp
+}
+
+func (s *PushAPIServer) getKeepTimestamp() bool {
+	s.rwMutex.RLock()
+	defer s.rwMutex.RUnlock()
+	return s.keepTimestamp
+}
+
+func (s *PushAPIServer) SetRelabelRules(rules frelabel.Rules) {
+	s.rwMutex.Lock()
+	defer s.rwMutex.Unlock()
+	s.relabelRules = frelabel.ComponentToPromRelabelConfigs(rules)
+}
+
+func (s *PushAPIServer) getRelabelRules() []*relabel.Config {
+	s.rwMutex.RLock()
+	defer s.rwMutex.RUnlock()
+	newRules := make([]*relabel.Config, len(s.relabelRules))
+	for i, r := range s.relabelRules {
+		var rCopy = *r
+		newRules[i] = &rCopy
+	}
+	return newRules
+}
```

Comment:
(nit, not a blocker) I'm not super convinced about the added complexity yet; in truth I don't imagine it will be often that people change the relabel rules or labels by hand that we'd win anything over not having to restart the server every time, and needing the setter/getter/having to copy the fields approach etc.